equal_test.go

```diff
@@ -0,0 +1,166 @@
+	assertTrue(t, testAlmostEqual(math.Inf(-1), math.Inf(-1)))
+	assertFalse(t, testAlmostEqual(math.Inf(-1), math.Inf(1)))
+	assertFalse(t, testAlmostEqual(math.Inf(1), math.MaxFloat64))
+	assertFalse(t, testAlmostEqual(math.Inf(-1), -math.MaxFloat64))
+}
+
+func TestAlmostEqual_nan(t *testing.T) {
+	// Comparisons involving NaN values.
+	assertFalse(t, testAlmostEqual(math.NaN(), math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), 0.0))
+	assertFalse(t, testAlmostEqual(-0.0, math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), -0.0))
+	assertFalse(t, testAlmostEqual(0.0, math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), math.Inf(1)))
+	assertFalse(t, testAlmostEqual(math.Inf(1), math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), math.Inf(-1)))
+	assertFalse(t, testAlmostEqual(math.Inf(-1), math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), math.MaxFloat64))
+	assertFalse(t, testAlmostEqual(math.MaxFloat64, math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), -math.MaxFloat64))
+	assertFalse(t, testAlmostEqual(-math.MaxFloat64, math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), math.SmallestNonzeroFloat64))
+	assertFalse(t, testAlmostEqual(math.SmallestNonzeroFloat64, math.NaN()))
+	assertFalse(t, testAlmostEqual(math.NaN(), -math.SmallestNonzeroFloat64))
+	assertFalse(t, testAlmostEqual(-math.SmallestNonzeroFloat64, math.NaN()))
+}
+
+func TestAlmostEqual_opposite(t *testing.T) {
+	// Comparisons of numbers on opposite sides of 0
+	assertFalse(t, testAlmostEqual(1.000000001, -1.0))
+	assertFalse(t, testAlmostEqual(-1.0, 1.000000001))
+	assertFalse(t, testAlmostEqual(-1.000000001, 1.0))
+	assertFalse(t, testAlmostEqual(1.0, -1.000000001))
+	assertTrue(t, testAlmostEqual(10*math.SmallestNonzeroFloat64, 10*-math.SmallestNonzeroFloat64))
+	assertFalse(t, testAlmostEqual(10000*math.SmallestNonzeroFloat64, 10000*-math.SmallestNonzeroFloat64))
+}
+
+func TestAlmostEqual_ulp(t *testing.T) {
+	// The really tricky part - comparisons of numbers very close to zero.
+	assertTrue(t, testAlmostEqual(math.SmallestNonzeroFloat64, math.SmallestNonzeroFloat64))
+	assertTrue(t, testAlmostEqual(math.SmallestNonzeroFloat64, -math.SmallestNonzeroFloat64))
+	assertTrue(t, testAlmostEqual(-math.SmallestNonzeroFloat64, math.SmallestNonzeroFloat64))
+	assertTrue(t, testAlmostEqual(math.SmallestNonzeroFloat64, 0))
+	assertTrue(t, testAlmostEqual(0, math.SmallestNonzeroFloat64))
+	assertTrue(t, testAlmostEqual(-math.SmallestNonzeroFloat64, 0))
+	assertTrue(t, testAlmostEqual(0, -math.SmallestNonzeroFloat64))
+
+	assertFalse(t, testAlmostEqual(0.000000001, -math.SmallestNonzeroFloat64))
+	assertFalse(t, testAlmostEqual(0.000000001, math.SmallestNonzeroFloat64))
+	assertFalse(t, testAlmostEqual(math.SmallestNonzeroFloat64, 0.000000001))
+	assertFalse(t, testAlmostEqual(-math.SmallestNonzeroFloat64, 0.000000001))
+}
+
+func testAlmostEqual(a, b float64) bool {
+	return AlmostEqual(a, b, 0.00001)
```

Comment:
Just a style comment: Why not let `testAlmostEqual` take a `t` parameter and one about the expected outcome (true or false) and format a nice test log message in here? The code above would be simpler, `assertTrue` and `assertFalse` could go, and any failed test would describe much more nicely what went wrong.